lib/server/controllers/file_controller.rb

```diff
@@ -72,36 +74,116 @@ def rename
+
+    validate_copy_revision(revision)
+
+    new_file = Metis::File.copy({
+      project_name: @params[:project_name],
+      source_file: revision.source.file,
+      dest_file_path: revision.dest.file_path,
+      dest_bucket_name: revision.dest.bucket_name,
+      user: @user
+    })
 
-    raise Etna::Error.new('File not found', 404) unless file&.has_data?
+    success_json(files: [ new_file.to_hash(@request) ])
+  end
 
-    raise Etna::BadRequest, 'Invalid path' unless Metis::File.valid_file_path?(@params[:new_file_path])
+  def bulk_copy
+    # We want to support bulk copy operations
+    #   from Magma.
+    # First, check that the user has access to all of
+    #   the revision buckets and files, and that
+    #   all revisions are valid.
+    # If all revisions are valid, execute them.
+    require_param(:revisions)
 
-    raise Etna::Forbidden, "Cannot edit project #{@params[:project_name]}" unless @user.can_edit?(@params[:project_name])
+    revisions = @params[:revisions].
+      map {|rev| Metis::CopyRevision.new(rev) }
 
-    new_folder_path, new_file_name = Metis::File.path_parts(@params[:new_file_path])
+    raise Etna::BadRequest, 'At least one revision required' unless revisions.length > 0
+    raise Etna::BadRequest, 'All revisions require valid "source" parameter' unless revisions.all? { |rev| rev.source.valid?}
+    raise Etna::BadRequest, 'All revisions require valid "dest" parameter' unless revisions.all? { |rev| rev.dest.valid?}
 
-    new_bucket = require_bucket(@params[:new_bucket_name])
+    # In bulk copy mode, we validate the buckets in bulk
+    #   since they aren't part of the path (are part of revisions),
+    #   and trying to minimize database hits by rejecting early, if
+    #   possible.
+    all_source_bucket_names = revisions.map {|rev| rev.source.bucket_name}.uniq
+    all_dest_bucket_names = revisions.map {|rev| rev.dest.bucket_name}.uniq
 
-    new_folder = require_folder(new_bucket, new_folder_path)
+    hmac = @request.env['etna.hmac']
 
-    raise Etna::Forbidden, 'Folder is read-only' if new_folder&.read_only?
+    user_authorized_bucket_names = Metis::Bucket.where(
+      project_name: @params[:project_name],
+      owner: ['metis', hmac.id.to_s],
+      name: all_source_bucket_names + all_dest_bucket_names
+    ).all.select{|b| b.allowed?(@user, hmac)}.
+      map {|bucket| bucket.name}
+
+    raise Etna::Forbidden, 'User does not have access to all source buckets' unless revisions.all? {
+      |rev| rev.valid?('source_bucket_access', user_authorized_bucket_names)}
+    raise Etna::Forbidden, 'User does not have access to all dest buckets' unless revisions.all? {
+      |rev| rev.valid?('dest_bucket_access', user_authorized_bucket_names)}
+
+    revisions.each do |revision|
+      validate_copy_revision(revision)
+    end
+
+    # If we've gotten here, every revision looks good and we can execute them!
+    new_files = []
```

Comment:
You might consider moving this operation to the revision object also, in which case you could do: `success_json(files: revisions.map(&:revise!))`

Reply:
Good idea!